Cam sima dycore update

PeterHjortLauritzen opened the pull request:
Tag name (required for release branches):
Originator(s):

Description (include the issue title, and the keyword ['closes', 'fixes', 'resolves'] followed by the issue number):

This PR should enable the use of the SE dycore with the CAM4 physics suite. 

Describe any changes made to build system:

M       cime_config/cam_autogen.py
 - Prevent metadata file clobbering.

Describe any changes made to the namelist:

M       cime_config/namelist_definition_cam.xml
 - Add "scale_dry_air_mass" namelist entry.

M       src/dynamics/se/namelist_definition_se_dycore.xml
 - Update SE dycore namelist entries.

List any changes to the defaults for the input datasets (e.g. boundary datasets):

List all files eliminated and why:

D       src/dynamics/se/test_fvm_mapping.F90
 - Testing infrastructure no longer used/needed.

List all files added and what they do:

A       cime_config/testdefs/testmods_dirs/cam/outfrq_se_cam4/shell_commands
A       cime_config/testdefs/testmods_dirs/cam/outfrq_se_cam4/user_nl_cam
 - Add SE dycore-specific output to regression test


List all existing files that have been modified, and describe the changes: 
(Helpful git command: `git diff --name-status development...<your_branch_name>`)

M       src/dynamics/se/advect_tend.F90
M       src/dynamics/se/dp_coupling.F90
M       src/dynamics/se/dp_mapping.F90
M       src/dynamics/se/dycore/comp_ctr_vol_around_gll_pts.F90
M       src/dynamics/se/dycore/control_mod.F90
M       src/dynamics/se/dycore/dimensions_mod.F90
M       src/dynamics/se/dycore/edge_mod.F90
M       src/dynamics/se/dycore/element_mod.F90
M       src/dynamics/se/dycore/fvm_consistent_se_cslam.F90
M       src/dynamics/se/dycore/fvm_mapping.F90
M       src/dynamics/se/dycore/fvm_mod.F90
M       src/dynamics/se/dycore/fvm_reconstruction_mod.F90
M       src/dynamics/se/dycore/global_norms_mod.F90
M       src/dynamics/se/dycore/hybrid_mod.F90
M       src/dynamics/se/dycore/namelist_mod.F90
M       src/dynamics/se/dycore/prim_advance_mod.F90
M       src/dynamics/se/dycore/prim_advection_mod.F90
M       src/dynamics/se/dycore/prim_driver_mod.F90
M       src/dynamics/se/dycore/prim_init.F90
M       src/dynamics/se/dycore/prim_state_mod.F90
M       src/dynamics/se/dycore/quadrature_mod.F90
M       src/dynamics/se/dycore/viscosity_mod.F90
M       src/dynamics/se/dyn_comp.F90
M       src/dynamics/se/dyn_grid.F90
M       src/dynamics/se/gravity_waves_sources.F90
M       src/dynamics/se/namelist_definition_se_dycore.xml
M       src/dynamics/se/native_mapping.F90
M       src/dynamics/se/stepon.F90
 - SE dycore updates from CAM.

R089    src/dynamics/se/dycore/time_mod.F90     src/dynamics/se/dycore/se_dyn_time_mod.F90
 - Rename SE dycore source code file to avoid name collision.

M       src/control/cam_initfiles.F90
 - Implement dry air mass scaling.

M       src/control/runtime_opts.F90
- Remove un-used constituents namelist read.

M       src/data/air_composition.F90
 - Only use advected constituents in air composition calculation.

M       src/data/registry.xml
 - Modify 'dycore_calculates_geopotential_using_logarithms' variable to be False for SE dycore, but true for FV3 dycore.

M       src/data/registry_v1_0.xsd
 - Allow for numbers in dycore string (i.e. to use "FV3" as an option).
 
M       src/dynamics/mpas/dyn_comp_impl.F90
 - Remove un-used "readtrace" logic when reading in constituent values.

M       src/dynamics/utils/dynconst.F90
 - Pass standard pressure (pstd) to dycore.

M       src/history/cam_history.F90
 -  Add "is_history_field_active" function, which is needed by the SE dycore diagnostics.

M       src/history/cam_history_support.F90
 - Code clean-up.

M       src/physics/utils/cam_constituents.F90
 - Remove "cam_constituents_readnl" subroutine, and add new "const_diag_name" function to allow for the querying of a constituents "diagnostic_name" property.

M       src/utils/cam_abortutils.F90
 - Adjust left the allocation error message string.

M       src/utils/cam_map_utils.F90
 -  Update map size error message to provide the actual (problematic) values.

M       cime_config/testdefs/testlist_cam.xml
M       cime_config/testdefs/testmods_dirs/cam/outfrq_analy_ic_cam4/user_nl_cam
 - Add new QPC4 SE dycore regression tests and add extra comments to namelist file.

If there are new failures (compared to the `test/existing-test-failures.txt` file),
have them OK'd by the gatekeeper, note them here, and add them to the file.
If there are baseline differences, include the test and the reason for the
diff. What is the nature of the change? Roundoff?

derecho/intel/aux_sima:

All tests have expected NLCOMP failures due to new "scale_dry_air_mass" namelist entry, as well as namelist changes to the SE dycore (which do not impact FPHYStest runs).

SMS_D_Ln9.ne3pg3_ne3pg3_mt232.QPC4.derecho_intel.cam-outfrq_se_cam4 (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mt232.QPC4.derecho_intel.cam-outfrq_se_cam4 (Overall: DIFF)
 - New tests (so no currently archived baselines)

SMS_Ln9.ne3pg3_ne3pg3_mg37.FADIAB.derecho_intel.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FCAM7.derecho_intel.cam-outfrq_se_cslam_analy_ic (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FHS94.derecho_intel.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FKESSLER.derecho_intel.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FTJ16.derecho_intel.cam-outfrq_se_cslam (Overall: DIFF)
 - Answer differences due to changes in SE dycore.

SMS_Ln9.ne3pg3_ne3pg3_mg37.FKESSLER.derecho_intel.cam-outfrq_se_cslam_multitape (Overall: DIFF)
 - Known failure

derecho/gnu/aux_sima:

All tests have expected NLCOMP failures due to new "scale_dry_air_mass" namelist entry, as well as namelist changes to the SE dycore (which do not impact FPHYStest runs).

SMS_D_Ln9.ne3pg3_ne3pg3_mt232.QPC4.derecho_gnu.cam-outfrq_se_cam4 (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mt232.QPC4.derecho_gnu.cam-outfrq_se_cam4 (Overall: DIFF)
 - New tests (so no currently archived baselines)

SMS_Ln9.ne3pg3_ne3pg3_mg37.FADIAB.derecho_gnu.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FCAM7.derecho_gnu.cam-outfrq_se_cslam_analy_ic (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FHS94.derecho_gnu.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FKESSLER.derecho_gnu.cam-outfrq_se_cslam (Overall: DIFF)
SMS_Ln9.ne3pg3_ne3pg3_mg37.FTJ16.derecho_gnu.cam-outfrq_se_cslam (Overall: DIFF)
 - Answer differences due to changes in SE dycore.

If this changes climate describe any run(s) done to evaluate the new
climate in enough detail that it(they) could be reproduced:

CAM-SIMA date used for the baseline comparison tests if different than latest:

Mjoldnir committed:
remove lcp_moist and line-by-line merge of subroutine vertical_remap

Mjoldnir committed:
forgot to remove lcp_moist from dp_coupling

Mjoldnir committed:
remove phys_dyn_cp, Rayleigh friction (incl. namelist), vertical diffusion, add se_molecular_diff namelist

Mjoldnir committed:
rename nu_s->nu_t

Mjoldnir committed:
remove unused variables (element_mod.F90 merged)

Mjoldnir committed:
merge fvm_consistent.F90 (efficiency updates from Jim Edwards for Derecho)

Mjoldnir committed:
merge use_cslam logic (instead of ntrac>0)

Mjoldnir committed:
merge fvm_reconstruction_mod

Mjoldnir committed:
first extensive merge with cam_development

jimmielin committed:
Merge branch 'development' into hplin/dycore_update_from_peter

jimmielin committed:
Fix nested timer ESMF error

PeterHjortLauritzen committed:
Merge pull request #1 from jimmielin/hplin/dycore_update_from_peter

Merge in check_energy & cam_thermo_water_update changes

PeterHjortLauritzen committed:
remove debug statement

PeterHjortLauritzen committed:
change namelist defaults

jimmielin reviewed (commented):
Thanks @PeterHjortLauritzen! Some early comments.

I think there are a handful of instances in the dycore that have code walled off using `#ifdef 0` and a comment, for much of the code that depends on tracers/chemistry/constituents and history, which were previously unsupported. Not sure if we should begin to implement these now or continue to wall these off (perhaps in a constituent manner to what is there now, i.e., change all to `#ifdef 0`) so they can be dealt with together at a later time.

PeterHjortLauritzen committed:
performance enhancements from https://github.com/ESCOMP/CAM/pull/1365

PeterHjortLauritzen committed:
minor fixes

jimmielin commented on cime_config/namelist_definition_cam.xml:
```diff
@@ -189,7 +189,27 @@
+      dry surface pressure scaling. If less than or equal to 0.0,
+      do not perform scaling. If greater than 0.0, perform scaling to scale_dry_air_mass
+      value (in Pa) as the average dry surface pressure target.
+      Default: set by build-namelist.
```
@nusbaume might have thoughts about this but I generally just left this line out when moving over the description from CAM

jimmielin reviewed (approved):
Thanks @PeterHjortLauritzen!

Since this is bit-for-bit with cam_development and I have no further questions about the code I'll approve it from my end! Depending on whether this or the gravity waves PR goes in first I'm happy to handle whatever merge conflicts arise due to vorticity code updates in the GW PR that also affect the SE dycore.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Make modifications found during code review.

nusbaume committed:
Additional clean-up found during code review.

nusbaume committed:
Another round of code review cleanup.

nusbaume committed:
Yet another round of code review cleanup.

nusbaume committed:
Final round of code review cleanup.

nusbaume committed:
Bring in changes from CAM tag cam6_4_150, plus minval->maxval bug fix.

nusbaume committed:
Bring in changes from CAM tag cam6_4_154

nusbaume committed:
Bring in changes from CAM tag cam6_4_155

nusbaume committed:
Update history and constituents code for 'dyn_comp' and 'stepon'.

nusbaume committed:
Remove un-used testing code, and perform additional cleanup.

nusbaume committed:
Implement final cleanup and history/constituent implementation steps.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Fix build errors.

nusbaume committed:
Use constituent diagnostic names instead of standard names in history fields.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Implement changes needed for SE dycore to run with CAM4 and real ICs.

nusbaume committed:
Remove unused 'readtrace' variable and cam_constituents 'readnl' subroutine.

nusbaume committed:
Add new grid and looping structure to output 'FX' variables on GLL grid.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

jimmielin committed:
Proposed fix for statediag_numtrac

Assisted-by: claude-opus:4.6

nusbaume committed:
Add aquaplanet CAM4 regression tests with SE dycore.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Fix comments and atm log output.

nusbaume committed:
Merge pull request #2 from jimmielin/CAM-SIMA-dycore-update

Proposed fix for statediag_numtrac

nusbaume committed:
Fix dbuf3 bug, and bring in latest atmospheric_physics tag.

nusbaume committed:
Merge remote-tracking branch 'refs/remotes/origin/CAM-SIMA-dycore-update' into CAM-SIMA-dycore-update

nusbaume committed:
Fix bug preventing SIMA from using metadata file in SourceMods.

Note that CAM-SIMA currently still requires the source (e.g. F90)
file to be present in SourceMods as well, but at least now metadata
files will be seen in the case where both are present.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Make SE dycore robust against tracer order in constituents object, and fix bug in geopotential_temp call.

Commit:
Bit-for-bit fixes for SE dycore update

Commit:
Bit-for-bit fixes for SE-CSLAM

Commit:
Fix memory leaks and mismatched constituent/advected tracer indices.

Commit:
Remove debug printout

Commit:
Fix stale dbuf3 buffer

PeterHjortLauritzen committed:
Merge pull request #3 from jimmielin/CAM-SIMA-dycore-update

Bit-for-bit fixes to match CAM

jimmielin commented on src/dynamics/se/dp_coupling.F90:
```diff
@@ -762,20 +769,20 @@ subroutine derived_phys_dry(cam_runtime_opts, phys_state, phys_tend)
-   call qneg_run('D_P_COUPLING', columns_on_task, pver, &
-                 qmin_vals, const_data_ptr, errflg, errmsg)
+   call qneg_run('D_P_COUPLING', pcols, pver, qmin_vals, &
+                 const_data_ptr, errflg, errmsg)
```
I think we might want to move this qneg block below the dry to wet conversion to be consistent with CAM:

https://github.com/ESCOMP/CAM/blob/5f582d41d2a109a5eb5369e0c8b7975d8e2ea4c4/src/dynamics/se/dp_coupling.F90#L719-L735

nusbaume replied:
I've moved the `qneg_run` call down to match CAM.  Now even if something bad happens it will be the same badness that CAM has!

jimmielin commented on src/dynamics/se/namelist_definition_se_dycore.xml:
```diff
@@ -318,45 +256,64 @@
-    <category>se</category>
+  <entry id="se_sponge_del4_nu_div_fac">
+    <type>real</type>
+    <category>se></category>
```
```suggestion
    <category>se</category>
```

nusbaume replied:
Good catch!  Extra `>` removed.

jimmielin commented on src/dynamics/se/namelist_definition_se_dycore.xml:
```diff
@@ -318,45 +256,64 @@
+  </entry>
+  <entry id="se_sponge_del4_lev">
+    <type>real</type>
+    <category>se></category>
```
```suggestion
    <category>se</category>
````

nusbaume replied:
Thanks!  Extra `>` removed.

jimmielin commented on src/dynamics/se/namelist_definition_se_dycore.xml:
```diff
@@ -318,45 +256,64 @@
+    </values>
+  </entry>
+  <entry id="se_sponge_del4_lev">
+    <type>real</type>
```
Should this be `integer` to be consistent with CAM?

nusbaume replied:
Yes, good find!  I've updated the type to be `integer` (which is what it should be for both CAM and SIMA).

jimmielin commented on src/dynamics/se/dycore/prim_advection_mod.F90:
```diff
@@ -1112,14 +1101,15 @@ subroutine vertical_remap(hybrid,elem,fvm,hvcoord,np1,np1_qdp,nets,nete)
           call omp_set_nested(.true.)
-          !$OMP PARALLEL NUM_THREADS(tracer_num_threads), DEFAULT(SHARED), PRIVATE(hybridnew2,qbeg,qend)
+          !$OMP PARALLEL NUM_THREADS(tracer_num_threads),
+          !DEFAULT(SHARED), PRIVATE(hybridnew2,qbeg,qend)
```
```suggestion
          !$OMP PARALLEL NUM_THREADS(tracer_num_threads) &
          !$OMP DEFAULT(SHARED) &
          !$OMP PRIVATE(hybridnew2,qbeg,qend)
```

nusbaume replied:
Wow, did you catch this on your own, or with Claude's help?  Either way it was a great find, as the syntax was definitely wrong.  OpenMP directives updated as suggested.

jimmielin replied:
I saw this one because it's a stray diff but definitely couldn't spot the stray > in the namelist files without Claude!

jimmielin reviewed (approved):
Thanks @nusbaume! I just had a few minor suggested changes but I am happy with the rest of the code since I've ran this a couple of times already and knowing this is bfb with CAM in FADIAB. I will bring in the IC fixes after this PR is merged.

nusbaume committed:
Make modifications found during code review.

nusbaume committed:
Merge remote-tracking branch 'upstream/development' into CAM-SIMA-dycore-update

nusbaume committed:
Remove no longer failing test from existing-test-failures.txt.
